### ISSUE TITLE
Wazuh v3.7.1 manager processes crash issue on ubuntu 16.04

### ISSUE THREAD

**Issue by whatthejay** [wazuh/wazuh#2088]
## Bug report

OS: Ubuntu 16.04
Wazuh version: 3.7.1
Wazuh API version: 3.7.1
Install type: manager
Install method: packages (apt)
Log sample
===========================================================
# dmesg log

[277056.851495] ossec-remoted[17809]: segfault at 7fd7610df1c0 ip 00007fd7610df1c0 sp 00007ffc1dafcb78 error 14
[277058.289980] ossec-analysisd[17885]: segfault at 7f5691f0b260 ip 00007f5691f0b260 sp 00007f795cff8e90 error 14
[277066.745865] wazuh-db[18074]: segfault at 1 ip 000000000040430f sp 00007f18c16cdf08 error 6 in wazuh-db[400000+31000]
[277077.182117] wazuh-modulesd[18480]: segfault at 0 ip 0000000000404a50 sp 00007f8c7fffe178 error 6 in wazuh-modulesd[400000+7d000]
[277077.555171] wazuh-db[18420]: segfault at 0 ip 00000000004042f2 sp 00007f88b4313f08 error 4 in wazuh-db[400000+31000]
[277080.820181] ossec-analysisd[18490]: segfault at 116 ip 0000000000405fd0 sp 00007f817b2d5f28 error 6 in ossec-analysisd[400000+b5000]
[277338.723144] wazuh-db[18755]: segfault at 20 ip 0000000000419160 sp 00007f2937397e88 error 4 in wazuh-db[400000+31000]
[277359.393706] wazuh-db[23134]: segfault at 1 ip 000000000040430a sp 00007fd9dada9f08 error 4 in wazuh-db[400000+31000]
[277374.743882] traps: ossec-analysisd[23790] trap invalid opcode ip:461153 sp:7f5912ffcf00 error:0 in ossec-analysisd[400000+b5000]
[277437.750121] traps: ossec-analysisd[25944] trap invalid opcode ip:461153 sp:7f69907efea0 error:0 in ossec-analysisd[400000+b5000]
[281048.937210] wazuh-db[26340]: segfault at 1 ip 0000000000404309 sp 00007f71ec05ff10 error 6 in wazuh-db[400000+31000]
[284650.665251] wazuh-db[30061]: segfault at 0 ip 00000000004042f0 sp 00007fad7a424f10 error 6 in wazuh-db[400000+31000]
[287904.434638] ossec-analysisd[26418]: segfault at ffffffffbe1b79c3 ip 0000000000405590 sp 00007f992ffdeca8 error 7 in ossec-analysisd[400000+b5000]
[288264.442577] wazuh-db[30364]: segfault at 1 ip 000000000040430f sp 00007f2c171d8f08 error 6 in wazuh-db[400000+31000]
[295112.306417] ossec-analysisd[30733]: segfault at 0 ip 0000000000413ee0 sp 00007f62d97f1ea8 error 6 in ossec-analysisd[400000+b5000]
[295464.199643] ossec-remoted[31379]: segfault at 7fe09b399f88 ip 00007fe09b192a25 sp 00007febfe7dbcb0 error 4
[298710.215837] ossec-analysisd[31438]: segfault at 7f9a85216870 ip 00007f9a85216870 sp 00007fe0767ebe88 error 14
[302307.706797] wazuh-db[30807]: segfault at 7f7acce061e0 ip 00007f7acce061e0 sp 00007f3c3e71fe88 error 14
[302663.163311] wazuh-db[32111]: segfault at 1 ip 000000000040430f sp 00007f85ae035f08 error 6 in wazuh-db[400000+31000]
[305913.152856] wazuh-db[32209]: segfault at 1 ip 000000000040430f sp 00007febf97c7f08 error 6 in wazuh-db[400000+31000]
[306266.580155] wazuh-db[32482]: segfault at 786b74 ip 000000000045ab50 sp 00007fa6d8e2f930 error 4 in wazuh-db[630000+1000]
[309507.761161] wazuh-db[32575]: segfault at 1 ip 000000000040430f sp 00007faa4e142f08 error 6 in wazuh-db[400000+31000]
[316706.672541] wazuh-db[381]: segfault at 786b74 ip 000000000045ab50 sp 00007f3c267ca930 error 4 in wazuh-db[630000+1000]
[316712.099126] ossec-analysisd[31847]: segfault at 7fa9fe0ed260 ip 00007fa9fe0ed260 sp 00007f7aa2ffce90 error 14
[320304.261174] ossec-analysisd[1050]: segfault at 0 ip 0000000000413ee0 sp 00007fbdfffeeea8 error 6 in ossec-analysisd[400000+b5000]
[320321.544163] ossec-analysisd[1417]: segfault at 0 ip 0000000000413ee0 sp 00007f44997f1ea8 error 6 in ossec-analysisd[400000+b5000]
[322851.846593] ossec-logcollec[26439]: segfault at 0 ip 0000000000408ec3 sp 00007f3d5a1c7f10 error 6 in ossec-logcollector[400000+86000]
[323909.730128] wazuh-db[908]: segfault at 1 ip 000000000040430f sp 00007fa9fa017f08 error 6 in wazuh-db[400000+31000]
[327513.723106] ossec-analysisd[1612]: segfault at 19d ip 0000000000405fd0 sp 00007fe3801f9f28 error 6 in ossec-analysisd[400000+b5000]

# /var/ossec/logs/ossec.log

2018/12/13 06:37:18 wazuh-modulesd:syscollector: INFO: Starting evaluation.
2018/12/13 06:37:18 ossec-analysisd: ERROR: at sc_send_db(): Connection with wazuh-db lost. Reconnecting.
2018/12/13 06:37:18 ossec-analysisd: ERROR: at sc_send_db(): Connection with wazuh-db lost. Reconnecting.
2018/12/13 06:37:26 wazuh-modulesd:syscollector: INFO: Evaluation finished.
2018/12/13 06:39:58 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Ubuntu Xenial database update...
2018/12/13 06:46:44 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Ubuntu Trusty database update...
2018/12/13 06:54:41 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Debian Stretch database update...
2018/12/13 07:00:45 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Debian Jessie database update...
2018/12/13 07:07:35 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Debian Wheezy database update...
2018/12/13 07:07:55 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Red Hat Enterprise Linux 5 database update...
2018/12/13 07:07:57 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Red Hat Enterprise Linux 6 database update...
2018/12/13 07:07:58 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Red Hat Enterprise Linux 7 database update...
2018/12/13 07:07:59 wazuh-modulesd:vulnerability-detector: INFO: (5452): Starting vulnerability scanning.
2018/12/13 07:08:08 wazuh-modulesd:vulnerability-detector: INFO: (5453): Vulnerability scanning finished.
2018/12/13 07:13:09 wazuh-modulesd:vulnerability-detector: INFO: (5452): Starting vulnerability scanning.
2018/12/13 07:13:09 wazuh-modulesd:vulnerability-detector: INFO: (5453): Vulnerability scanning finished.
2018/12/13 07:18:10 wazuh-modulesd:vulnerability-detector: INFO: (5452): Starting vulnerability scanning.
2018/12/13 07:18:10 wazuh-modulesd:vulnerability-detector: INFO: (5453): Vulnerability scanning finished.
2018/12/13 07:23:11 wazuh-modulesd:vulnerability-detector: INFO: (5452): Starting vulnerability scanning.
2018/12/13 07:23:11 wazuh-modulesd:vulnerability-detector: INFO: (5453): Vulnerability scanning finished.
2018/12/13 07:28:12 wazuh-modulesd:vulnerability-detector: INFO: (5452): Starting vulnerability scanning.
2018/12/13 07:28:12 wazuh-modulesd:vulnerability-detector: INFO: (5453): Vulnerability scanning finished.
2018/12/13 07:33:13 wazuh-modulesd:vulnerability-detector: INFO: (5452): Starting vulnerability scanning.
2018/12/13 07:33:13 wazuh-modulesd:vulnerability-detector: INFO: (5453): Vulnerability scanning finished.
2018/12/13 07:37:03 ossec-remoted: ERROR: socketerr (not available).
2018/12/13 07:37:03 ossec-remoted: ERROR: (1210): Queue '/queue/ossec/queue' not accessible: 'Connection refused'.
2018/12/13 07:37:03 ossec-remoted: ERROR: socketerr (not available).
2018/12/13 07:37:03 ossec-remoted: ERROR: (1210): Queue '/queue/ossec/queue' not accessible: 'Bad file descriptor'.
2018/12/13 07:37:03 ossec-remoted: ERROR: socketerr (not available).
2018/12/13 07:37:03 ossec-remoted: ERROR: (1210): Queue '/queue/ossec/queue' not accessible: 'Bad file descriptor'.
2018/12/13 07:37:03 ossec-remoted: ERROR: socketerr (not available).
2018/12/13 07:37:03 ossec-remoted: ERROR: (1210): Queue '/queue/ossec/queue' not accessible: 'Bad file descriptor'.
2018/12/13 07:37:03 ossec-logcollector: ERROR: socketerr (not available).
2018/12/13 07:37:04 ossec-logcollector: ERROR: socketerr (not available).

===========================================================

Bug description
----------------
Some processes in the manager host are *very frequently* crashed.
Crashed processes are usually ossec-analysisd, ossec-remoted, wazuh-db.
I have no clue why the processes are being crashed.
I'm running **monit** as a watchdog.

When I restart all processes with */var/ossec/bin/ossec-control restart*, then all services work normally, but after some time (the time varies, 10sec ~ more than 1 hour) the processes are found dead.

Please let me know what I have to do to find out the reason and fix it.

**Comment by chemamartinez** [comment]
Hi @whatthejay,

Thank you for reporting this bug.

Actually, yesterday we found an issue in `ossec-analysisd` that make it crash when Windows agents with Wazuh v3.7.0 are connected to a manager v3.7.1 and its username contains spaces. I was wondering if it could be your case.

That issue has already fixed at #2089.

Apart from that, it could be interesting to know more about your environment to try to find out the cause of the crash. Could you share us the following information?

- The manager configuration (`ossec.conf`).
- Number of agents connected.
- Resources of the manager system, mostly the memory RAM.

Regards,
Chema.

**Comment by whatthejay** [comment]
@chemamartinez Thank you for your interest.

1. Attached manager configuration (/var/ossec/etc/ossec.conf)
2. We have currently 3 agents connected: 1 ubuntu, 1 centos, 1 windows server
3. Wazuh is running on OpenStack KVM machine with 8 vCPUs, 16GB RAM, 1TB SSD Disk

Best Regards,
Jay.

[ossec.conf.zip](https://github.com/wazuh/wazuh/files/2678793/ossec.conf.zip)

**Comment by whatthejay** [comment]
> Hi Jay,
> 
> Everything seems normal in your configuration and resources. I was wondering if you could help us to find out the crash source by following the next steps.
> 
> 1. Download a debug supported version of the `ossec-analysisd` binary [here](https://github.com/wazuh/wazuh/files/2686122/ossec-analysisd.zip).
> 2. Stop the manager by running `service wazuh-manager stop`.
> 3. Replace the downloaded binary of `ossec-analysisd` by the one located at `/var/ossec/bin`. I suggest you back up the original one before.
> 4. Once the binary has been overwritten, download a package called `apport` used to generate crash reports in Ubuntu. You just have to run this command:
> 
> ```
> apt-get install -y apport
> ```
> 1. Ensure the directory `/var/crash` exists on your manager host. If not, create it.
> 2. Restart the manager (`service wazuh-manager start`), and when it crashes again, check that it has been generated the following file: `/var/crash/_var_ossec_bin_ossec-analysisd.0.crash`.
> 3. Create a temporary folder, `/root/crash-tmp` for example, and unpack the crash report with `apport` as follow:
> 
> ```
> apport-unpack /var/crash/_var_ossec_bin_ossec-analysisd.0.crash /root/crash-tmp
> ```
> 1. Finally, send us the crash report as well as the content of the `crash-tmp` folder, in a ZIP file:
> 
> ```
> cp -f /var/crash/_var_ossec_bin_ossec-analysisd.0.crash /root/crash-tmp
> zip -r crash.zip /root/crash-tmp
> ```
> Analyzing that report, we hope to find the cause of the crash and fix it as soon as possible.
> 
> Best regards,
> Chema.

Dear @chemamartinez 

I *accidentally* upgraded the manager to 3.7.2, and the problem still exists.
I found /var/crash already exists, so I've tried *apport-unpack* but failed with the following exception.

```
Traceback (most recent call last):
  File "/usr/bin/apport-unpack", line 73, in <module>
    pr.extract_keys(f, bin_keys, dir)
  File "/usr/lib/python3/dist-packages/problem_report.py", line 270, in extract_keys
    [item for item, element in b64_block.items() if element is False])
ValueError: ['UserGroups'] has no binary content
```

So, could you please upload 3.7.2 version of debug-enabled binary for "ossec-analysisd" and "wazuh-db"? Then I'll upload the crash report.

Thank you.

**Comment by chemamartinez** [comment]
Hi @whatthejay, 

Yes, of course, please find the both requested binaries [here](https://github.com/wazuh/wazuh/files/2689464/daemons.zip)

That python exception is a known problem, it happens to me as well. However, the report is unpacked correctly even though the exception.

Please, before running the manager again, ensure to delete the content of the folder `/var/crash` to be sure the report sent is the new one.

Regards,
Chema.

**Comment by whatthejay** [comment]
Hi, @chemamartinez,

This is the crash report for ossec-analysisd.

Regards,
Jay.

[crash_ossec_analysisd.zip](https://github.com/wazuh/wazuh/files/2689585/crash_ossec_analysisd.zip)

**Comment by chemamartinez** [comment]
Hi @whatthejay,

I have been analyzing the report with the team. We have not obtained a clear conclusion about the source of the issue. We know it is related to a stack memory issue, probably particular on OpenStack KVM.

If you could share us information about the user limits of your machine, it could be helpful for clarifying the cause. It is enough with the output of the command `ulimit -a`.

I will try to reproduce the crash on an OpenStack KVM, to be able to find a solution.

Regards,
Chema.

**Comment by whatthejay** [comment]
@chemamartinez,
Here's the result of ulimt -a (run with root)

```
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 128625
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 128625
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

Regards,
Jay.

**Comment by V1D1AN** [comment]
I have the same problem with the Wazuh 3.7.1 ( branch fix-windows-dec )

**Comment by whatthejay** [comment]
@chemamartinez,
 I've installed v3.7.2 on CentOS 7 in same OpenStack environment.
It works with no problem. Just FYI.

**Comment by iasdeoupxe** [comment]
Strange enough it seems the same (at least the log file entries reported in the first post) started to show up in my Debian 9 environment after upgrading the wazuh-manager from 3.7.2 to 3.8.0 (no crashes / issues previously with 3.7.2 and below).

I will try to gather more information like previously described to see if those are really the same. But at least ossec-remoted seems to be the culprit in my environment and is crashing.

**Comment by vikman90** [comment]
Hi, @whatthejay @V1D1AN @iasdeoupxe,

This issue seems to be fixed since 3.7.2. Let me close it. Please reopen it if the problem appears again.

Thank you.
Best regards.
